python/pip_install/tools/wheel_installer/wheel.py

```diff
@@ -245,26 +264,33 @@ def _add(self, dep: str, platform: Optional[Platform]):
         if dep == self.name:
             return
 
-        if platform:
-            self._select[platform].add(dep)
-        else:
+        if not platform:
             self._deps.add(dep)
+            return
 
-    @staticmethod
-    def _normalize(name: str) -> str:
-        return re.sub(r"[-_.]+", "_", name).lower()
+        # Add the platform-specific dep
+        self._select[platform].add(dep)
 
-    def add(self, *wheel_reqs: str) -> None:
-        reqs = [Requirement(wheel_req) for wheel_req in wheel_reqs]
+        # Add the dep to more specific constraints
+        for p in platform:
+            if p not in self._select:
+                continue
 
-        # Resolve any extra extras due to self-edges
-        self._want_extras = self._resolve_extras(reqs)
+            self._select[p].add(dep)
```

Comment:
The `if p not in self._select` part of this block of code is confusing me.

The iteration will produce one of two list of values:
1. Exactly one element: self (may or may not have arch set)
2. Two or more elements: self (without an arch) and one or more with arch set.

In case one, L279 duplicates what L272 did: the loop is one element long, p is present, and L279 executes, and we're done.

In case two, the same thing happens: the loop is 1 + N long. The first element does L279 (duplicating L272), and then subsequent iterations do nothing (the arch-bound platform isn't present).

What am I missing? That some earlier add_req() call added an entry? But then wouldn't that mean some of these were ignored when they shouldn't have been...? or maybe the loop below has something to do with this? I think if I could see the requirements lines being processed it'd be easier for me to reason about.

Reply:
Added a unit-test and more comments to hopefully make the code more self-explanatory.